sharedbuffers/mapped_struct.py

```diff
@@ -2059,33 +2064,433 @@ def hinted_bsearch(a, hkey, hint):
+                while pindex2 < pend2 and cython.cast(cython.p_float, pindex2)[0] <= ref and pdest < pdestend:
+                    cython.cast(cython.p_float, pdest)[0] = cython.cast(cython.p_float, pindex2)[0]
+                    cython.cast(cython.p_float, pdest)[1] = cython.cast(cython.p_float, pindex2)[1]
+                    pdest += stride0
+                    pindex2 += stride0
+
+        # Copy leftover tails
+        while pindex1 < pend1 and pdest < pdestend:
+            cython.cast(cython.p_float, pdest)[0] = cython.cast(cython.p_float, pindex1)[0]
+            cython.cast(cython.p_float, pdest)[1] = cython.cast(cython.p_float, pindex1)[1]
+            pdest += stride0
+            pindex1 += stride0
+        while pindex2 < pend2 and pdest < pdestend:
+            cython.cast(cython.p_float, pdest)[0] = cython.cast(cython.p_float, pindex2)[0]
+            cython.cast(cython.p_float, pdest)[1] = cython.cast(cython.p_float, pindex2)[1]
+            pdest += stride0
+            pindex2 += stride0
+        return (pdest - pdeststart) / stride0
 
+    # Commented cython directives in pxd
     #@cython.ccall
-    @cython.locals(lo = cython.size_t, hi = cython.size_t, ix = cython.size_t, hint = cython.size_t)
-    #@cython.returns(cython.bint)
-    def hinted_sorted_contains(a, hkey, hint):
-        hi = len(a)
-        ix = hinted_bsearch(a, hkey, hint)
-        if ix >= hi:
-            return False
-        else:
-            return a[ix] == hkey
+    @cython.locals(
+        stride0 = cython.size_t, length1 = cython.size_t, length2 = cython.size_t,
+        destlength = cython.size_t, rdestlength = cython.size_t,
+        index1buf = 'Py_buffer', index2buf = 'Py_buffer', destbuf = 'Py_buffer',
+        pindex1 = cython.p_char, pindex2 = cython.p_char, pdest = cython.p_char)
+    #@cython.returns(cython.size_t)
+    def index_merge(index1, index2, dest):
+        length1 = len(index1)
+        length2 = len(index2)
+        destlength = len(dest)
+        rdestlength = min(destlength, length1+length2)
+
+        if len(index1.shape) != 2 or len(index2.shape) != 2:
+            raise ValueError("Indices must be two-dimensional")
+        if len(dest.shape) != 2:
+            raise ValueError("Destination must be two-dimensional")
+        if index1.shape[1] != 2 or index2.shape[1] != 2:
+            raise ValueError("Indices must be N x 2 matrices")
+        if dest.shape[1] != 2:
+            raise ValueError("Destination must be an N x 2 matrix")
 
-    #@cython.ccall
-    @cython.locals(lo = cython.size_t, hi = cython.size_t)
-    #@cython.returns(cython.bint)
-    def sorted_contains(a, hkey):
-        hi = len(a)
-        lo = 0
-        return hinted_sorted_contains(a, hkey, (lo+hi)//2)
+        #lint:disable
+        PyObject_GetBuffer(index1, cython.address(index1buf), PyBUF_STRIDED_RO)
+        try:
+            PyObject_GetBuffer(index2, cython.address(index2buf), PyBUF_STRIDED_RO)
+            try:
```

Comment:
Nope, each try is independent, if the second or third initialization fail, the ones that succeeded have to be properly released, that's why there are nested try/finally blocks. They cannot be unified without a lot of spaghetti.